### DISCUSSION TITLE
Override keymaps for prompt windows

### DISCUSSION THREAD

**Discussion by mystilleef** [olimorris/codecompanion.nvim#533] [Q&A] [answered]
`CodeCompanion` distinguishes between hiding and closing windows. Closing chat windows destroys them. 

This is great for ephemeral windows like the commit message prompt window. But for the main chat window hiding is the preferable option as they're intended to be toggled. 

As the Lord intended, I've bound `q` and `esc` to hide chat windows. But I want those same keys to destroy certain prompt windows. I've tried to get used to `<c-c>`, but this old brain is not having it. I've also looked into events. There's a `closed` chat event, but none for `hidden`. 

Is there a way to override the default keymaps for prompt windows?

**Comment by olimorris** [comment]
You could add your own custom keymap. Below is a non-related example that I've used previously (source: #485)

```lua
require('codecompanion').setup({
  strategies = {
    chat = {
      keymaps = {
        hide = {
          modes = {
            n = "gh",
          },
          callback = function(chat)
            chat.ui:hide()
          end,
          description = "Hide the chat buffer",
        },
      }
    }
  }
})
```

You could take this principle and customize it to suit your needs.

**Reply by mystilleef** [reply]
Thanks for your quick response. Yes, but those keymaps apply to all chat windows. I was wondering if there was a way to override the keymaps for some prompt-library windows.

**Reply by olimorris** [reply]
Ahhhhh. I see what you mean. I could always create a custom autocmd when a prompt library is triggered?

**Reply by mystilleef** [reply]
Yeah, that sounds good. For now, I've found a hacky workaround that works. Since I turn off system prompts for prompt-library windows, I use `chat.opts.ignore_system_prompt` to decide whether to close or hide the window. Ideally, an `opts.keymaps` table should be in the prompt-library settings that override the global chat settings.

**Reply by olimorris** [reply, marked as answer]
I've just added a new event called `CodeCompanionChatOpened` which a boolean value `from_prompt_library`.

**Reply by mystilleef** [reply]
Thank you! I now use `chat.from_prompt_library` to decide whether to close or hide chat windows.
